Formal2024/Workshops/workshop3.lean

```diff
@@ -0,0 +1,210 @@
+/-
+  # Logic: Conjunction, Iff, and Disjunction
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+import Mathlib.Tactic
+
+/-!
+  # Conjunction
+-/
+
+-- The symbol for conjunction is `∧`, which you can write as `\and`.
+
+-- To prove a conjunction, we use the `constructor` tactic.
+-- This will split the goal into two, one for each part of the conjunction.
+-- (For people familiar with Lean 3: the `split` tactic is gone in Lean 4.)
+
+example (P Q : Prop) (hp : P) (hq : Q) : P ∧ Q := by
+  constructor
+  · exact hp
+  · exact hq
+
+-- Here is another example:
+example {x y : ℝ} (h₀ : x ≤ y) (h₁ : ¬y ≤ x) : x ≤ y ∧ x ≠ y := by
+  constructor
+  · assumption -- This tactic looks for a hypothesis in the context that matches the goal, like `h₀`.
+  · intro h
+    apply h₁
+    rw [h]
+
+-- You can also use the anonymous constructor `⟨⟩`:
+example {x y : ℝ} (h₀ : x ≤ y) (h₁ : ¬y ≤ x) : x ≤ y ∧ x ≠ y := by
+  have h : x ≠ y := by
+    contrapose! h₁
+    rw [h₁]
+  exact ⟨h₀, h⟩
+
+
+-- To use a conjunction, you can use `rcases` or `rintro`, as with existential quantifiers.
+example {x y : ℝ} (h : x ≤ y ∧ x ≠ y) : ¬y ≤ x := by
+  rcases h with ⟨h₀, h₁⟩
+  contrapose! h₁
+  exact le_antisymm h₀ h₁
+
+example {x y : ℝ} : x ≤ y ∧ x ≠ y → ¬y ≤ x := by
+  rintro ⟨h₀, h₁⟩ h'
+  exact h₁ (le_antisymm h₀ h')
+
+-- You can also access the parts of a conjunction `h` with `h.left` and `h.right` (or `h.1` and `h.2`).
+example {x y : ℝ} (h : x ≤ y ∧ x ≠ y) : ¬y ≤ x :=
+  fun h' ↦ h.right (le_antisymm h.left h')
+
+-- You can next anonymous constructors, even if they mix conjunctions and existential quantifiers.
+-- Note that they (usually) associate to the right.
+example : ∃ x : ℕ, 3 < x ∧ x < 5 ∧ ∃ y, y^2 = x :=
+  ⟨4, by norm_num, by norm_num, 2, by norm_num⟩
+
+
+/-!
+  # Iff
+-/
+
+-- The symbol for iff is `↔`, which you can write as `\iff`.
+
+-- As with `∧`, you can use the `constructor` tactic, or the anonymous constructor `⟨⟩` to prove an iff.
+-- Similarly, you can use an iff with `rcases` or `rintro`.
+-- If `h` is an iff, you can access the two implications with `h.mp` and `h.mpr` (or `h.1` and `h.2`).
+
+example {x y : ℝ} (h : x ≤ y) : ¬y ≤ x ↔ x ≠ y := by
+  constructor
+  · intro h' heq
+    apply h'
+    rw [heq]
+  · intro h' hle
+    apply h'
+    exact le_antisymm h hle
+
+
+-- Lean knows that `↔` is reflexive, symmetric, and transitive, so you can use the `rw` tactic with it.
+#check abs_lt
+
+example (x : ℝ) : |x + 3| < 5 → -8 < x ∧ x < 2 := by
+  rw [abs_lt]
+  intro h
+  constructor <;> linarith
+-- The `<;>` tactic combinator attempts to close every open goal with the given tactic
+
```

Comment:
Maybe "all the open goals opened by constructor" is more precise? I would also add a remark that "<" is different from the anonymus constructor "\<"